alea/include/alps/alea/serialize.hpp

```diff
@@ -6,186 +6,65 @@
-template <typename T>
-T scalar_deserialize(deserializer &ser, const std::string &key)
-{
-    T value;
-    ser.read(key, ndview<T>(&value, nullptr, 0));
-    return value;
-}
-
-/** Helper function for deserialization of scalars */
-template <typename T>
-void scalar_deserialize(deserializer &ser, const std::string &key, T &value)
-{
-    ser.read(key, ndview<T>(&value, nullptr, 0));
-}
-
-}}}
-
-namespace alps { namespace alea {
-
-// Serialization methods
-
-inline void serialize(serializer &ser, const std::string &key, uint32_t value) {
-    internal::scalar_serialize(ser, key, value);
-}
-inline void serialize(serializer &ser, const std::string &key, int32_t value) {
-    internal::scalar_serialize(ser, key, value);
-}
-inline void serialize(serializer &ser, const std::string &key, uint64_t value) {
-    internal::scalar_serialize(ser, key, value);
-}
-inline void serialize(serializer &ser, const std::string &key, int64_t value) {
-    internal::scalar_serialize(ser, key, value);
-}
-inline void serialize(serializer &ser, const std::string &key, double value) {
-    internal::scalar_serialize(ser, key, value);
-}
-inline void serialize(serializer &ser, const std::string &key,
-                      std::complex<double> value) {
-    internal::scalar_serialize(ser, key, value);
-}
+namespace alps { namespace serialization {
 
+/** Serializes Eigen array of complex_op<double> */
 template <typename Derived>
-void serialize(serializer &ser, const std::string &key,
-               const Eigen::MatrixBase<Derived> &value)
+typename std::enable_if<
+        eigen_scalar_is<Derived, alps::alea::complex_op<double>>::value>::type
+serialize(serializer &ser, const std::string &key,
+          const Eigen::PlainObjectBase<Derived> &value)
 {
-    typedef Eigen::internal::traits<Derived> traits;
-    typedef typename traits::Scalar scalar_type;
-    typedef Eigen::Matrix<scalar_type, Derived::RowsAtCompileTime,
-                          Derived::ColsAtCompileTime> plain_matrix_type;
+    using scalar_type = eigen_scalar_t<Derived>;
+    using matrix_type = Eigen::Matrix<scalar_type, Eigen::Dynamic, Eigen::Dynamic>;
 
-    // Ensure that evaluated expression will be continuous
-    if ((Derived::MaxRowsAtCompileTime != Eigen::Dynamic
-                    && Derived::MaxRowsAtCompileTime != value.rows())
-            || (Derived::MaxColsAtCompileTime != Eigen::Dynamic
-                    && Derived::MaxColsAtCompileTime != value.cols())
-            || ((Derived::Options & Eigen::RowMajor)
-                    && value.rows() != 1 && value.cols() != 1))
-        serialize(ser, key, plain_matrix_type(value));
+    // Ensure that evaluated expression will be Fortran-contiguous
+    if (!eigen_is_contiguous(value))
+        serialize(ser, key, matrix_type(value));
 
     // Evaluate to matrix or proxy object if already matrix
-    auto temp = value.eval();
-
+    const double *dbl_data = reinterpret_cast<const double *>(value.data());
```

Comment:
This `reinterpret_cast` (and another one below) looked scary enough for me to spend some time on cppreference and SE, but i think i got it finally. :) Is it to deal with complex numbers? 
In this case, it could be nice to add a comment here, some time down the road -- no need to delay the merge because of this.

Reply:
You're absolutely right ... it's to deal with `complex_op` entities, which are essentially 2x2 matrices, laid out in memory like `double[2][2]`.